### PR TITLE
guardrails: restrict replication factor (RF)

### LINKED ISSUE #8619: Add option to restrict Replication Factor
On a multi-rack deployment (e.g., X nodes per Rack/AZ), it is strongly advised to use a RF to match the number of racks.
In such a case, it will be useful for the system admin, who manages the deployment, to limit the replication strategy and factor to values that make sense, in this case, multiplication of X.

### LINKED ISSUE #8892: Idea: add an option for default replication strategy and RF for new keyspaces
Today, when creating a keyspace a replication strategy and replication factor must be explicitly stated - for example:
```
CREATE KEYSPACE ks WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 3 }
```
This syntax is verbose and hard to remember, which too often leads users to blindly copy example they find on the web - and often the examples they find are bad - such as using `SimpleStrategy` (see #8586) or using an ill-advised RF (see #8891, #8619).

If we allow configuring (separately) a default replication class and a default replication factor, it will be harder to mistakenly use a bad one.

Cassandra has an issue https://issues.apache.org/jira/browse/CASSANDRA-14557 about adding a `default_keyspace_rf` configuration option (there is also an uncommitted patch - https://github.com/sumanth-pasupuleti/cassandra/commit/f9fb9537e2c6a41b3df13c7c4af2d23b0912b18c). I'm not aware of an idea to add a default replication _strategy_ in Cassandra.

These defaults can be especially useful for managed deployments like Scylla Cloud, where by virtue of the number of nodes, DCs and racks, there is a recommended replication strategy and factor that the user should most probably use - unless they really know what they are doing. See also #8619.

### PR DESCRIPTION
**Pull request by ptrsmrn**
Replacing `minimum_keyspace_rf` config option with 4 config options:
`{minimum,maximum}_replication_factor_{warn,fail}_threshold`, which
allow us to impose soft limits (issue a warning) and hard limits (not
execute CQL) on RF when creating/altering a keyspace.
The reason to rather replace than extend `minimum_keyspace_rf` config
option is to be aligned with Cassandra, which did the same, and has the
same parameters' names.
Only min soft limit is enabled by default and it is set to 3, which means
that we'll generate a CQL warning whenever RF is set to either 1 or 2.
RF's value of 0 is always allowed and means that there will not be any
replicas on a given DC. This was agreed with PM.
Because we don't allow to change guardrails' values when scylla is
running (per PM), there're no tests provided with this PR, and dtests will be
provided separately.
Exceeding guardrails' thresholds will be tracked by metrics.

Resolves https://github.com/scylladb/scylladb/issues/8619
Refs https://github.com/scylladb/scylladb/issues/8892 (the RF part, not the replication-strategy part)

### CODE REVIEW

**Comment by mykaul** [comment]
Will you be adding tests to the PR?

**Comment by nyh** [comment]
>  The reason to rather replace than extend `minimum_keyspace_rf` config option is to be aligned with Cassandra, which did the same, and has the same parameters' names.

Please note that Cassandra only has the "minimum" pair of options - it doesn't have the maximum. I still don't think anyone needs the "maximum" limits.

File: cql3/statements/create_keyspace_statement.cc
Comment (nyh) [inline cql3/statements/create_keyspace_statement.cc:194]: Nitpick: the min_fail>=0 check isn't needed, because you already know rf > 0, so if min_fail wasn't >=0, you couldn't have rf < min_fail.
Since you don't need to check min_fail >=0 you also don't need to add the min_fail variable, which you only use once.

Comment (nyh) [reply cql3/statements/create_keyspace_statement.cc:194]: I guess with max_fail below the situation is different, though.

File: db/config.cc
Comment (nyh) [inline db/config.cc:989]: If we start doing MustRestart for guard rails, perhaps we need to go back and do it for all of them, not just this one.
I'm not sure I'm happy about this. I think @tzach is mistaken here - why is Scylla Cloud worried specifically about guardrails being changable via CQL? I would think there are much more "dangerous" options to change - people can disable hints, can turn on or off various protocols, and other strange changes, why are we worried only about guardrails? Maybe Scylla cloud needs to disable *most* configuration changes?

File: conf/scylla.yaml
Comment (nyh) [inline conf/scylla.yaml]: I think you copied a comment from a completely unrelated test??

Comment (nyh) [reply conf/scylla.yaml]: Or, wait, what is this commit?
Maybe you should rebase...

**Comment by ptrsmrn** [comment]
> > The reason to rather replace than extend `minimum_keyspace_rf` config option is to be aligned with Cassandra, which did the same, and has the same parameters' names.
> 
> Please note that Cassandra only has the "minimum" pair of options - it doesn't have the maximum. I still don't think anyone needs the "maximum" limits.

Cassandra has both the min & max, at least on the latest `trunk` - https://github.com/apache/cassandra/blob/trunk/conf/cassandra.yaml#L1929
I think that max can make sense, though we're not using it. For example it could make sense to warn users about setting RF > number of nodes in DC.

File: conf/scylla.yaml
Comment (ptrsmrn) [reply conf/scylla.yaml]: This was auto-rebase commit made by github, which actually turned out to be a merge commit.

File: db/config.cc
Comment (ptrsmrn) [reply db/config.cc:989]: I thought we agreed these can be set only in the config file, the discussion has been reopened, and we're waiting for a final decision.

File: cql3/statements/create_keyspace_statement.cc
Comment (ptrsmrn) [reply cql3/statements/create_keyspace_statement.cc:194]: Yeah, but because guardrails with values lower than 0 are disabled, for consistency reasons I added everywhere a check that they're >= 0.

**Comment by ptrsmrn** [comment]
> Will you be adding tests to the PR?

@mykaul yes, in the end I decided to go with dtests, as they're flexible enough to cover various scenarios we need to test in order to ensure guardrails are really working (like setting a multi DC cluster, upgrade node, modify scylla.yaml, etc.), so there's going to be a separate PR opened in scylla-dtest git repo once we merge this PR.
Just to be clear, I want ALL the tests for guardrails to be in written in dtests, no good reason to scatter them across different testing frameworks.

**Comment by ptrsmrn** [comment]
These would be first dtests for (RF) guardrails - https://github.com/scylladb/scylla-dtest/pull/3251

CC @mykaul

File: db/config.cc
Comment (ptrsmrn) [reply db/config.cc:989]: > If we start doing MustRestart for guard rails, perhaps we need to go back and do it for all of them, not just this one.

Not sure this is required, since all of them are disabled (and I don't see any our clients using them).

> I thought we agreed these can be set only in the config file, the discussion has been reopened, and we're waiting for a final decision.

Since there's no resolution, I think I can change it back to `LiveUpdate` - that shouldn't hurt any of our client, be it on premise or cloud. We can still change it later.
So can we proceed with this PR?

Comment (avikivity) [reply db/config.cc:989]: LiveUpdate is not about CQL. It's about not having to restart the server.

Comment (ptrsmrn) [reply db/config.cc:989]: Right, because there's also an option to send `SIGHUP` to reload the `scylla.yaml` config.

**Comment by ptrsmrn** [comment]
@nyh ping

File: cql3/query_processor.cc
Comment (nyh) [reply cql3/query_processor.cc:459]: In general, if you decided now to add a counter for each guardrail, I think we should have an organized place to put all these counters together (but please correct me if I'm wrong).

File: db/config.cc
Comment (nyh) [reply db/config.cc:1006]: I still don't understand the MustRestart change.

**Review by nyh — COMMENTED** [review]
@ptrsmrn I can't approve or disapprove this patch. The code looks fine (I left just a couple of minor comments) but was also fine before this patch. This patch is almost 100% policy - why do we want or need these 4 configuration options instead of 1? why do want that MustRestart thing? Do we want counters for each guardrail violation?  @tzach and @eliransin I think you guys should review this patch and make the decision - and I'll do the merging once it is decided.

**Comment by nyh** [comment]
> @ptrsmrn I can't approve or disapprove this patch. The code looks fine (I left just a couple of minor comments) but was also fine before this patch. This patch is almost 100% policy - why do we want or need these 4 configuration options instead of 1? why do want that MustRestart thing?

@mykaul @eliransin in general, I think we need to start encouraging more people to review patches. I see more and more patches that remind me of Isidor Isaac Rabi's famous quote "Who Ordered That?" - clearly, somebody "ordered" these patches, somebody is waiting for these patches, somebody was in a discussion that encouraged work in this direction - so it would be nice if that somebody can also review the result - they don't need to be maintainers to do such reviews.

File: conf/scylla.yaml
Comment (ptrsmrn) [reply conf/scylla.yaml:596]: I added a github issue to do the actual refactor - https://github.com/scylladb/scylladb/issues/14544 (note it's tracked under a general guardrails umbrella issue).

File: db/config.cc
Comment (ptrsmrn) [reply db/config.cc]: Given https://github.com/scylladb/scylladb/pull/14382, I think we're good with `liveness::LiveUpdate` now.

File: cql3/query_processor.cc
Comment (ptrsmrn) [reply cql3/query_processor.cc]: Changed names to `{minimum,maximum}_replication_factor_{warn,fail}_violations`.
My original idea was to keep the names exactly the same so it's easier to locate all the changes related to a given guardrail in the source code, but I agree the readability from user side is more important.
As for the groupings, or other refactor, if anything has to be done in this matter, I'd do it in https://github.com/scylladb/scylladb/issues/14544.

File: cql3/query_processor.cc
Comment (ptrsmrn) [reply cql3/query_processor.cc]: Changed as suggested.

File: cql3/statements/create_keyspace_statement.cc
Comment (eliransin) [inline cql3/statements/create_keyspace_statement.cc:175]: I know it is a nitpick however, there is a function better suited for this, `topology.get_datacenters` since you already refactored this code, it might be beneficial to make it more readable.

**Review by eliransin — APPROVED** [review]
LGTM

**Comment by ptrsmrn** [comment]
@nyh 
given that:

1. @eliransin approved this PR already (from the policy perspective)
2. Failing dtests are fixed in a PR in dtests repo - https://github.com/scylladb/scylla-dtest/pull/3251 
3. https://github.com/scylladb/scylladb/pull/14382, which allows for these guardrails to have `liveness::LiveUpdate` set, has been merged

I think we can merge this PR.

**Comment by nyh** [comment]
The CI failed. It's too late for me to check why this happened, I'm trying to restart the CI maybe it will pass this time.

**Comment by nyh** [comment]
@ptrsmrn please look into the CI failures. There are real failures in dtest, for example in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3228/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlsh/Tests___Sanity_Tests___test_batch/ you have a dtest complaining about the new warning you added:

```
>       assert 0 == len(stdout), stdout
E       AssertionError: 
E         Warnings :
E         Using Replication Factor replication_factor=1 lower than the minimum_replication_factor_warn_threshold=3 is not recommended.
```

**Comment by ptrsmrn** [comment]
@nyh yes, the are failures in dtests, which I have a PR in dtests repo for. I mentioned it in [my comment](https://github.com/scylladb/scylladb/pull/14262#issuecomment-1689042204), sorry if that wasn't clear.

**Comment by nyh** [comment]
> @nyh yes, the are failures in dtests, which I have a PR in dtests repo for. I mentioned it in [my comment](https://github.com/scylladb/scylladb/pull/14262#issuecomment-1689042204), sorry if that wasn't clear.

I didn't realize that that dtest patch contains both: 1. Fixes for existing dtests to make them pass with your code, and 2. New tests for your code. But I don't see how we can have both changes in the same dtest PR - when can we commit this change? We can't commit it before the Scylla code fix (because the new test will fail) and we can't commit it after (because the Scylla code fix will fail with the unmodified dtest).

**Comment by ptrsmrn** [comment]
@nyh the CI is finally green, I guess we're good to merge this PR.

**Comment by nyh** [comment]
> @nyh the CI is finally green, I guess we're good to merge this PR.

Done. Thanks.

**Pull request closed without merging**
